### ISSUE TITLE
YOLO12 models training collapsed for segment and classify tasks

### ISSUE THREAD

**Issue by Laughing-q** [ultralytics/ultralytics#19414]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

@ultralytics/yolo @sunsmarterjie I was using the same training parameters and training settings as YOLO11 to train YOLO12 segment and classify models. However I encountered the issue that some of the trainings collapsed.
- Segment
  - YOLO12N/L/X-seg: training goes well for now
  - YOLO12S-seg: training collapsed
- Classify
  - YOLO12N-cls: training collapsed

The training command for these two collapsed ones:
```bash
# segment
yolo segment train model=yolo12s-seg.yaml data=coco.yaml epochs=600 batch=128 project=YOLO12 name=yolo12s-seg cache=disk scale=0.9 copy_paste=0.15 mixup=0.05 device=1

# classify
yolo classify train device=2 model=yolo12n-cls.yaml data=../datasets/imagenet epochs=200 batch=256 imgsz=224 project=YOLO12-cls name=yolo12n-cls lr0=0.2 weight_decay=1e-4 warmup_epochs=0 nbs=256 hsv_s=0.4 erasing=0.4 momentum=0.9 auto_augment=randaugment cos_lr=True optimizer=SGD workers=16
```

Related training dashboard: 
- Segment: https://wandb.ai/laughing/YOLO12
- Classify: https://wandb.ai/laughing/YOLO12-cls

@

### Environment

```bash
Ultralytics 8.3.78 🚀 Python-3.11.10 torch-2.5.0+cu124 CUDA:0 (NVIDIA A100-SXM4-80GB, 81038MiB)
Setup complete ✅ (128 CPUs, 1007.7 GB RAM, 3133.3/3519.3 GB disk)

OS                  Linux-5.15.0-122-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.11.10
Install             git
RAM                 1007.71 GB
Disk                3133.3/3519.3 GB
CPU                 AMD EPYC 7502 32-Core Processor
CPU count           128
GPU                 NVIDIA A100-SXM4-80GB, 81038MiB
GPU count           8
CUDA                12.4

numpy               ✅ 1.26.1>=1.23.0
numpy               ✅ 1.26.1<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.0+cu124>=1.8.0
torch               ✅ 2.5.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.0+cu124>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.10>=2.0.0
```

### Minimal Reproducible Example

```bash
# segment
yolo segment train model=yolo12s-seg.yaml data=coco.yaml epochs=600 batch=128 project=YOLO12 name=yolo12s-seg cache=disk scale=0.9 copy_paste=0.15 mixup=0.05 device=1

# classify
yolo classify train device=2 model=yolo12n-cls.yaml data=../datasets/imagenet epochs=200 batch=256 imgsz=224 project=YOLO12-cls name=yolo12n-cls lr0=0.2 weight_decay=1e-4 warmup_epochs=0 nbs=256 hsv_s=0.4 erasing=0.4 momentum=0.9 auto_augment=randaugment cos_lr=True optimizer=SGD workers=16
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Laughing-q, thank you for bringing this up and for providing detailed information 🚀! We recommend reviewing the [Docs](https://docs.ultralytics.com/) for guidance, particularly the [Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) that may help optimize your YOLO12 training.

If this is a 🐛 Bug Report, a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) is essential to help us investigate. From the details provided, you've already shared your commands for segment and classify tasks—thank you! If anything else can assist in reproducing the issue, such as logs or additional dataset specifics, feel free to share.

If this is a custom training ❓ Question, please ensure your training settings and data preparation align with the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Providing your unaltered YAML config and insights on any deviations from standard workflows will be invaluable.

To stay connected with the Ultralytics community and explore solutions collaboratively:
- Chat with us on [Discord](https://discord.com/invite/ultralytics) 🎧
- Participate in detailed discussions on [Discourse](https://community.ultralytics.com/)
- Share your ideas or ask questions on [Reddit](https://reddit.com/r/Ultralytics)

## Upgrade

Always ensure you're using the latest `ultralytics` package and dependencies. Update with the following command to verify if your issue persists in the most recent release:

```bash
pip install -U ultralytics
```

## Environments

We suggest testing your training in a verified and up-to-date environment to eliminate potential compatibility issues. Some of the supported setups include:
- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

The current state of the [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) can inform you about platform-wide verification of YOLO operations. If green, all core functionality for [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) is validated daily.

📢 This is an automated response to guide you through resources and steps while ensuring no essential details are overlooked. An Ultralytics engineer will review your case and provide personalized assistance soon 😊.

**Comment by Laughing-q** [comment]
@sunsmarterjie Do you think this is related to the `residual` connection? since with current trainings only small sizes of models are collapsing i.e. N/S. 
From the paper I noticed the team encountered the same issue with L/X sizes, which has been fixed by introducing `residual` connection to them, but does not implement the `residual` connection to small sizes like N/S/M. Do you think this might be the root of this? I've launched another YOLO12L-cls training to verify this theory, let's see how it goes.

**Comment by KuiperSirius** [comment]
Hi @Laughing-q :-)

Thank you for your detailed report and valuable contributions to YOLO12. I am currently encountering challenges in my training process due to GPU memory constraints, which limit my ability to experiment with more optimal parameters. Would you be willing to share the yolo12n-seg.pt pretrained weights based on the COCO-Seg dataset? I believe these weights could significantly enhance the rigor and reproducibility of my academic experimental evaluations.

Wishing you a swift resolution to the bug, and thank you again for your support and efforts.

Best regards~

**Comment by Laughing-q** [comment]
@KuiperSirius Hi there, the models are still training and we'll open source them once trainings are done!

**Comment by KuiperSirius** [comment]
Hi @Laughing-q   :-) Really appreciate your efforts!👍 🦖

**Comment by KuiperSirius** [comment]
@Laughing-q  🦖 Thank you again for your prompt response. I've been conducting further experiments on my side using your pre-training parameters for yolov12n-seg with a batch size of 128 (imgsz=640). Under these conditions, I observed that the accuracy curve requires quite a few epochs to converge and smooth out. However, when using a batch size of 32, the curve seems to exhibit a steeper increase in fewer epochs. 

Perhaps you could try this parameter setting while keeping other parameters—such as imgsz and augmentation—at their default values. Based on my preliminary observations, if we can achieve mAP50-95 (box) and mAP50-95 (mask) exceeding 38.9 and 32.0 respectively, the performance of our pre-trained model would already surpass that of the previous v11n-seg.

Thanks again for your hard work~

Best regards,

**Comment by sunsmarterjie** [comment]
I think the residual connections for smaller models can help.

**Comment by glenn-jocher** [comment]
Hi there! 😊 Thank you for sharing your observations on residual connections and batch size impacts. We're actively investigating architectural refinements for smaller YOLO12 variants, including residual connections as mentioned in the [YOLO12 key improvements](https://docs.ultralytics.com/models/yolo12/#key-improvements). These changes will be carefully validated through our standard benchmarking process before release. Stay tuned for updates in future releases! 🚀 

For those interested in current best practices, we recommend YOLO11 models which show more stable training characteristics across all scales.

**Comment by javahook** [comment]
I met the same training collapsed problem(val: nan) with yolo12m-cls.pt

**Comment by glenn-jocher** [comment]
Hi @javahook, thanks for confirming this issue with yolo12m-cls as well. We're seeing a pattern with training instability in smaller YOLO12 variants.

The instability is likely related to the residual connection implementation as @sunsmarterjie suggested. While these connections were initially designed to stabilize larger models, the smaller architectures may need similar adaptations. 

We're actively working on architectural refinements to improve YOLO12's training stability across all model sizes. Until these improvements are implemented, I'd recommend:

1. Trying smaller batch sizes as @KuiperSirius suggested
2. If stability is critical for your project, consider using YOLO11 models which demonstrate more consistent training behavior across different scales

We'll update the repository once we've validated solutions for these issues. Thanks for your patience and for reporting your experiences.

**Comment by Jordan-Pierce** [comment]
@glenn-jocher does yolo12(n,s,m,l,x) not have pre-trained weights for other tasks like classify, segment, etc?

**Comment by glenn-jocher** [comment]
Hi @Jordan-Pierce — not currently; YOLO12 ships Detect checkpoints only, and pretrained Segment/Classify/Pose/OBB weights aren’t published. For those tasks please use the more stable [Ultralytics YOLO11 models](https://docs.ultralytics.com/models/yolo11/), or train YOLO12 from the task YAMLs like `yolo12s-seg.yaml` / `yolo12n-cls.yaml`; if you need follow-up, open a new issue for visibility. ([github.com](https://github.com/ultralytics/ultralytics/issues/19352))

**Comment by Jordan-Pierce** [comment]
> Hi @Jordan-Pierce — not currently; YOLO12 ships Detect checkpoints only, and pretrained Segment/Classify/Pose/OBB weights aren’t published. For those tasks please use the more stable [Ultralytics YOLO11 models](https://docs.ultralytics.com/models/yolo11/), or train YOLO12 from the task YAMLs like `yolo12s-seg.yaml` / `yolo12n-cls.yaml`; if you need follow-up, open a new issue for visibility. ([github.com](https://github.com/ultralytics/ultralytics/issues/19352))

Nope that answered it, thanks!

**Comment by glenn-jocher** [comment]
Glad that answered it—if anything else comes up, please open a new issue for visibility and include your `yolo checks` output plus the full text traceback; for general tips see our [Troubleshooting Common YOLO Issues](https://docs.ultralytics.com/guides/yolo-common-issues/) guide.
